File: TinyScript/JAVA/src/parser/ast/Expr.java
package parser.ast;

import lexer.Token;
import parser.util.ExprHOF;
import parser.util.ParseException;
import parser.util.PeekTokenIterator;
import parser.util.PriorityTable;

public class Expr extends ASTNode {
    private static PriorityTable table = new PriorityTable();

    public Expr(ASTNode parent) {
        super(parent);
    }

    public Expr(ASTNode parent, ASTNodeTypes type, Token lexeme) {
        super(parent);
        this.type = type;
        this.label = lexeme.getValue();
        this.lexeme = lexeme;
    }

    // E(k) -> E(k) op(k) E(k+1) | E(k+1)
    // right:
    //    E(k) -> E(k+1) E_(k)
    //        var e = new Expr(); e.left = E(k+1); e.op = op(k); e.right = E(k+1) E_(k)
    //    E_(k) -> op(k) E(k+1) E_(k) | ε
    // 最高优先级处理：
    //      E(t) -> F E_(k) | U E_(k)
    //      E_(t) -> op(t) E(t) E_(t) | ε
    private static ASTNode E(ASTNode parent, int k, PeekTokenIterator it) throws ParseException {
        if (k < table.size() - 1) {
            return combine(parent, it, () -> E(parent, k + 1, it), () -> E_(parent, k, it));
        } else {
            return race(
                    it,
                    () -> combine(parent, it, () -> U(parent, it), () -> E_(parent, k, it)),
                    () -> combine(parent, it, () -> F(parent, it), () -> E_(parent, k, it))
            );
        }
    }

    private static ASTNode E_(ASTNode parent, int k, PeekTokenIterator it) throws ParseException {
        var token = it.peek();
        var value = token.getValue();

        if (table.get(k).indexOf(value) != -1) {
            var expr = new Expr(parent, ASTNodeTypes.BINARY_EXPR, it.nextMatch(value));
            expr.addChild(combine(parent, it,
                    () -> E(parent, k + 1, it),
                    () -> E_(parent, k, it)
            ));
            return expr;
        }
        return null;
    }

    private static ASTNode U(ASTNode parent, PeekTokenIterator it) throws ParseException {
        var token = it.peek();
        var value = token.getValue();
        ASTNode expr = null;

        if (value.equals("(")) {
            it.nextMatch("(");
            expr = E(parent, 0, it);
            it.nextMatch(")");
            return expr;
        } else if (value.equals("++") || value.equals("--") || value.equals("!")) {
            var t = it.peek();
            it.nextMatch(value);
            Expr unaryExpr = new Expr(parent, ASTNodeTypes.UNARY_EXPR, t);
            unaryExpr.addChild(E(unaryExpr, 0, it));
            return unaryExpr;
        }

        return null;
    }

    private static ASTNode F(ASTNode parent, PeekTokenIterator it) {
        var token = it.peek();
        if (token.isVariable()) {
            return new Variable(parent, it);
        } else {
            return new Scalar(parent, it);
        }
    }

    private static ASTNode combine(ASTNode parent, PeekTokenIterator it, ExprHOF aFunc, ExprHOF bFunc) throws ParseException {
        // 如果传入的值是 null 的情况
        var a = aFunc.hoc();
        if (a == null) {
            return it.hasNext() ? bFunc.hoc() : null;
        }

        var b = it.hasNext() ? bFunc.hoc() : null;
        if (b == null) {
            return a;
        }

        // 传入的值是正常的情况
        Expr expr = new Expr(parent, ASTNodeTypes.BINARY_EXPR, b.lexeme);
        // expr.type = ASTNodeTypes.BINARY_EXPR;
        // expr.lexeme = b.lexeme;
        // expr.label = b.label;
        expr.addChild(a);
        expr.addChild(b.getChild(0));
        return expr;
    }

    private static ASTNode race(PeekTokenIterator it, ExprHOF aFunc, ExprHOF bFunc) throws ParseException {
        if (!it.hasNext()) {
            return null;
        }

        var a = aFunc.hoc();
        if (a != null) {
            return a;
        }
        return bFunc.hoc();
    }

    public static ASTNode parse(PeekTokenIterator it) throws ParseException {
        return E(null, 0, it);
    }
}

File: TinyScript/JS/src/parser/ast/Expr.js
const ASTNode = require("./ASTNode")
const ASTNodeTypes = require('./ASTNodeTypes')
const Scalar = require('./Scalar')
const Variable = require('./Variable')
const table = require('../util/PriorityTalbe')

class Expr extends ASTNode {
  constructor(parent) {
    super(parent)
  }

  static fromToken(parent, type, token) {
    const expr = new Expr(parent)
    expr.label = token.getValue()
    expr.lexeme = token
    expr.type = type
    return expr
  }

  /**
   * left: E(k) -> E(k) op(k) E(k+1) | E(k+1)
   * right:
   *   E(k) -> E(k+1) E_(k)
   *     // const e = new Expr() e.left = E(k+1) e.right = E_(k).child(0)
   *     // combine
   *   E_(k) -> op(k) E(k+1) E_(k) | ε
   *     // race
   *
   *   // U -> (E) | ++E | --E
   *   E(t) -> F E_(t) | U E_(t)
   * 
   * @param {*} parent 
   * @param {*} it 
   */
  static parseExpr(parent, it) {
    return Expr.E(parent, it, 0)
  }


  static E(parent, it, k) {
    if (k < table.length - 1) {
      return Expr.combine(
        parent,
        it,
        () => Expr.E(parent, it, k + 1),
        () => Expr.E_(parent, it, k)
      )
    } else {
      return Expr.race(
        it,
        () => Expr.combine(
          parent,
          it,
          () => Expr.F(parent, it),
          () => Expr.E_(parent, it, k)
        ),
        () => Expr.combine(
          parent,
          it,
          () => Expr.U(parent, it),
          () => Expr.E_(parent, it, k)
        )
      )
    }
  }

  // E_(k) -> op(k) E(k+1) E_(k) | ε
  static E_(parent, it, k) {
    const token = it.peek()
    const value = token.getValue()

    if (table[k].indexOf(value) !== -1) {
      it.nextMatch(value)
      const expr = Expr.fromToken(parent, ASTNodeTypes.BINARY_EXPR, token)
      expr.addChild(
        Expr.combine(
          parent,
          it,
          () => Expr.E(parent, it, k + 1),
          () => Expr.E_(parent, it, k)
        )
      )

      return expr
    }

    return null
  }

  static U(parent, it) {
    const token = it.peek()
    const value = token.getValue()

    if (value === "(") {
      it.nextMatch("(")
      const expr = Expr.parseExpr(parent, it)
      it.nextMatch(")")
      return expr
    } else if (value === "++" || value === "--" || value === "!") {
      const t = it.peek()
      it.nextMatch(value)

      const expr = Expr.fromToken(ASTNodeTypes.UNARY_EXPR, t)
      expr.addChild(Expr.parseExpr(parent, it))
      return expr
    }
    return null
  }

  static F(parent, it) {
    const token = it.peek()
    if (token.isVariable()) {
      return new Variable(parent, it)
    } else {
      return new Scalar(parent, it)
    }
  }

  static combine(parent, it, funcA, funcB) {
    if (!it.hasNext()) {
      return null
    }

    // a b 为 null 的情况
    const a = funcA()
    if (a == null) {
      return it.hasNext() ? funcB() : null
    }
    const b = it.hasNext() ? funcB() : null
    if (b == null) {
      return a
    }

    // a b 不为 null 的情况
    const expr = Expr.fromToken(parent, ASTNodeTypes.BINARY_EXPR, b.lexeme)
    expr.addChild(a)
    expr.addChild(b.getChild(0))
    return expr
  }

  static race(it, funcA, funcB) {
    if (!it.hasNext()) {
      return null
    }

    const a = funcA()
    if (a == null) {
      return funcB()
    }
    return a
  }
}

module.exports = Expr